### PR TITLE
Ensure XslCompiledTransform works with TrimMode=link

### LINKED ISSUE #45393: XslCompiledTransform doesn't work when using TrimMode=link
See the original report of this issue: https://github.com/dotnet/runtime/issues/44995#issuecomment-732819444

Run the following code in a Blazor WASM published app. Or in a console app that is PublishTrimmed=true && TrimMode=link

```C#
        public static void UseXsl()
        {
            using (StringReader text = new StringReader(
    @"<?xml version=""1.0"" encoding=""UTF-8""?>
<catalog>
    <cd>
        <title>Empire Burlesque</title>
        <artist>Bob Dylan</artist>
        <country>USA</country>
        <company>Columbia</company>
        <price>10.90</price>
        <year>1985</year>
    </cd>
</catalog>"))
            {


                XPathDocument myXPathDoc = new XPathDocument(text);
                XslCompiledTransform myXslTrans = new XslCompiledTransform();
                string xmlStr = @"<?xml version=""1.0""?>
<xsl:stylesheet version=""1.0"" xmlns:xsl=""http://www.w3.org/1999/XSL/Transform"">
<xsl:template match=""/"">
<html>
<body>
<h2> My CD Collection</h2>
   <table border = ""1"">
      <tr bgcolor = ""#9acd32"">
         <th> Title </th>
         <th> Artist </th>
       </tr>

      <xsl:for-each select = ""catalog/cd"">
          <tr>
            <td><xsl:value-of select = ""title"" /></td>
            <td><xsl:value-of select = ""artist"" /></td>
          </tr>
        </xsl:for-each>
       </table>
     </body>
     </html>
   </xsl:template>
   </xsl:stylesheet>";
                using (StringReader str = new StringReader(xmlStr))
                using (var reader = System.Xml.XmlReader.Create(str))
                {
                    myXslTrans.Load(reader);
                    StringWriter myWriter = new StringWriter();
                    myXslTrans.Transform(myXPathDoc, null, myWriter);
                    Console.WriteLine(myWriter.GetStringBuilder().ToString());

                }
            }

        }
```

### Expected results

The code should work the same way it does in a non-trimmed app.

### Actual results

```
(unknown) Unhandled Exception:
window.console.error
(unknown) System.ArgumentNullException: ArgumentNull_Generic Arg_ParamName_Name, con
window.console.error
(unknown)    at System.Reflection.Emit.CustomAttributeBuilder.Initialize(ConstructorInfo con, Object[] constructorArgs, PropertyInfo[] namedProperties, Object[] propertyValues, FieldInfo[] namedFields, Object[] fieldValues)
window.console.error
(unknown)    at System.Reflection.Emit.CustomAttributeBuilder..ctor(ConstructorInfo con, Object[] constructorArgs)
window.console.error
(unknown)    at System.Xml.Xsl.IlGen.XmlILModule.CreateLREModule()
window.console.error
(unknown)    at System.Xml.Xsl.IlGen.XmlILModule..cctor()
window.console.error
(unknown) crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: TypeInitialization_Type, System.Xml.Xsl.IlGen.XmlILModule
System.TypeInitializationException: TypeInitialization_Type, System.Xml.Xsl.IlGen.XmlILModule
 ---> System.ArgumentNullException: ArgumentNull_Generic Arg_ParamName_Name, con
   at System.Reflection.Emit.CustomAttributeBuilder.Initialize(ConstructorInfo con, Object[] constructorArgs, PropertyInfo[] namedProperties, Object[] propertyValues, FieldInfo[] namedFields, Object[] fieldValues)
   at System.Reflection.Emit.CustomAttributeBuilder..ctor(ConstructorInfo con, Object[] constructorArgs)
   at System.Xml.Xsl.IlGen.XmlILModule.CreateLREModule()
   at System.Xml.Xsl.IlGen.XmlILModule..cctor()
   Exception_EndOfInnerExceptionStack
   at System.Xml.Xsl.XmlILGenerator.Generate(QilExpression query, TypeBuilder typeBldr)
   at System.Xml.Xsl.XslCompiledTransform.CompileQilToMsil(XsltSettings settings)
   at System.Xml.Xsl.XslCompiledTransform.LoadInternal(Object stylesheet, XsltSettings settings, XmlResolver stylesheetResolver)
   at System.Xml.Xsl.XslCompiledTransform.Load(XmlReader stylesheet)
   at Panasoft.Web.MyData.Client.Pages.Monitor.TraceView.OnParametersSetAsync()
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle)
```

It looks like the following code needs to be annotated correctly for the ILLinker to preserve the constructors correctly:

https://github.com/dotnet/runtime/blob/aebd598664e7d931eee19eb1779c0ace5d1c3600/src/libraries/System.Private.Xml/src/System/Xml/Xsl/IlGen/GenerateHelper.cs#L135-L159

Side-note: The ILLink warnings caught this as well:

https://github.com/dotnet/runtime/blob/aebd598664e7d931eee19eb1779c0ace5d1c3600/src/libraries/System.Private.Xml/src/ILLink/ILLink.Suppressions.xml#L130-L141

### PR DESCRIPTION
**Pull request by eerhardt**
Resolve ILLinker warnings in Xml.Xsl in order to make a basic Xslt scenario work.

Fix #45393

### CODE REVIEW

**Comment by eerhardt** [comment]
mono windows x64 Debug failure is #45524

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/IlGen/GenerateHelper.cs [hunk 1]
```diff
@@ -134,273 +152,285 @@ public XmlILStorageMethods(Type storageType)
...
-        public static readonly MethodInfo XPFollMergeNext = GetMethod(typeof(XPathFollowingMergeIterator), "MoveNext");
-        public static readonly MethodInfo XPPrecCreate = GetMethod(typeof(XPathPrecedingIterator), "Create");
-        public static readonly MethodInfo XPPrecNext = GetMethod(typeof(XPathPrecedingIterator), "MoveNext");
-        public static readonly MethodInfo XPPrecDOCreate = GetMethod(typeof(XPathPrecedingDocOrderIterator), "Create");
-        public static readonly MethodInfo XPPrecDONext = GetMethod(typeof(XPathPrecedingDocOrderIterator), "MoveNext");
-        public static readonly MethodInfo XPPrecMergeCreate = GetMethod(typeof(XPathPrecedingMergeIterator), "Create");
-        public static readonly MethodInfo XPPrecMergeNext = GetMethod(typeof(XPathPrecedingMergeIterator), "MoveNext");
+        public static readonly MethodInfo AncCreate = typeof(AncestorIterator).GetMethod("Create")!;
+        public static readonly MethodInfo AncNext = typeof(AncestorIterator).GetMethod("MoveNext")!;
+        public static readonly MethodInfo AncCurrent = typeof(AncestorIterator).GetMethod("get_Current")!;
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/IlGen/GenerateHelper.cs:173]: Note here I needed to add a new field `Current` for each Iterator type, so we could safely call `get_Current` without the linker trimming it away.

**Review by vitek-karas — APPROVED** [review]
Nice!

**Comment by eerhardt** [comment]
@krwq @buyaa-n - any thoughts?

File: src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Xml.XPath;
+using System.Xml.Xsl;
+
+class Program
+{
+    static int Main(string[] args)
```
Comment (buyaa-n) [inline src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs:11]: Just curious is this only way to test trimming? Cannot have Unit test? Is this test would run with CI build

File: src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+<h2> My CD Collection</h2>
+   <table border = ""1"">
+      <tr bgcolor = ""#9acd32"">
+         <th> Title </th>
+         <th> Artist </th>
+       </tr>
+
+      <xsl:for-each select = ""catalog/cd"">
+          <tr>
+            <td><xsl:value-of select = ""title"" /></td>
```
Comment (buyaa-n) [inline src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs:45]: Nit: I understand this test only for showing if trimming works, just wondering if the updates made in `XmlIlVisitor.cs` are covered in the unit tests somewhere

**Review by buyaa-n — APPROVED** [review]
Overall looks good to me

File: src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs:11]: The "TrimmingTests" are run during the CI build with these legs:

image

These programs are published, executed, and the infrastructure verifies the application returns exit code `100` - which means "passed".

See https://github.com/dotnet/designs/blob/main/accepted/2020/linking-libraries.md#testing for more information.

File: src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs (hunk 3 above)
Comment (eerhardt) [reply src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs:45]: It looks like there are a bunch of XslCompiledTransform tests here:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.Xml/tests/Xslt/XslCompiledTransformApi/XslCompiledTransform.cs

File: src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs (hunk 2 above)
Comment (buyaa-n) [reply src/libraries/System.Private.Xml/tests/TrimmingTests/XslCompiledTransformTests.cs:11]: Cool, thanks!

**Review by krwq — APPROVED** [review]
LGTM, please make sure to run outerloop tests

**Comment by eerhardt** [comment]
The only outerloop failures are in RegularExpressions.Tests. Merging.

**Pull request merged**